app/frontend/Pages/Manage/CreateExperience/CreateBlockContext.tsx

```diff
@@ -0,0 +1,281 @@
+  validateMultistepForm,
+} from './CreateMultistepForm/CreateMultistepForm';
+import {
+  buildPollPayload,
+  canPollOpenImmediately,
+  getDefaultPollState,
+  processPollBeforeSubmit,
+  validatePoll,
+} from './CreatePoll/CreatePoll';
+import {
+  buildQuestionPayload,
+  canQuestionOpenImmediately,
+  getDefaultQuestionState,
+  processQuestionBeforeSubmit,
+  validateQuestion,
+} from './CreateQuestion/CreateQuestion';
+
+const CreateBlockContext = createContext<CreateBlockContextValue | null>(null);
+
+export function useCreateBlockContext() {
+  const context = useContext(CreateBlockContext);
+  if (!context) {
+    throw new Error('useCreateBlockContext must be used within a CreateBlockProvider');
+  }
+  return context;
+}
+
+interface CreateBlockProviderProps {
+  children: ReactNode;
+  participants: ParticipantSummary[];
+  onClose: () => void;
+  onEndCurrentBlock: () => Promise<void>;
+  refetchExperience: () => Promise<void>;
+}
+
+// NOTE: There are N number of branches for each block type. This is a good
+// candidate for a factory style pattern, but for now it is all centralized here
+// so we can keep adding too it without the conditional expansion leaking
+export function CreateBlockProvider({
+  children,
+  participants,
+  onClose,
+  onEndCurrentBlock,
+  refetchExperience,
+}: CreateBlockProviderProps) {
+  const [kind, setKind] = useState<Block['kind']>('poll');
+
+  const getDefaultState = useCallback((blockKind: Block['kind']): BlockData => {
+    switch (blockKind) {
+      case 'poll':
+        return getDefaultPollState();
+      case 'question':
+        return getDefaultQuestionState();
+      case 'multistep_form':
+        return getDefaultMultistepFormState();
+      case 'announcement':
+        return getDefaultAnnouncementState();
+      case 'mad_lib':
+        return getDefaultMadLibState();
+      default:
+        throw new Error(`Unknown block kind: ${blockKind}`);
+    }
+  }, []);
+
+  const [data, setData] = useState<BlockData>(() => getDefaultState(kind));
+
+  const [visibleRoles, setVisibleRoles] = useState<string[]>([]);
+  const [visibleSegmentsText, setVisibleSegmentsText] = useState<string>('');
+  const [targetUserIdsText, setTargetUserIdsText] = useState<string>('');
+  const [viewAdditionalDetails, setViewAdditionalDetails] = useState<boolean>(false);
+
+  const {
+    createExperienceBlock,
+    isLoading: isSubmitting,
+    error: createError,
+    setError: setCreateError,
+  } = useCreateExperienceBlock({ refetchExperience });
+
+  const handleKindChange = useCallback(
+    (newKind: Block['kind']) => {
+      setKind(newKind);
+      setData(getDefaultState(newKind));
+    },
+    [getDefaultState],
+  );
+
+  const submit = useCallback(
+    async (status: BlockStatus) => {
+      setCreateError(null);
+
+      let validationError: string | null = null;
+
+      switch (kind) {
+        case 'poll':
+          validationError = validatePoll(data as any);
+          break;
```

Comment:
Can we do `data.kind` in order to get type safety? Alternatively each one of these validation functions can update their type to just `BlockData` and do their own `kind` checks within

```suggestion
      switch (data.kind) {
        case 'poll':
          validationError = validatePoll(data);
          break;
```